packages/@aws-cdk/toolkit-lib/lib/api/drift/drift.ts

```diff
@@ -0,0 +1,87 @@
+import { format } from 'util';
+import type { DescribeStackDriftDetectionStatusCommandOutput, DescribeStackResourceDriftsCommandOutput } from '@aws-sdk/client-cloudformation';
+import { ToolkitError } from '../../toolkit/toolkit-error';
+import type { ICloudFormationClient } from '../aws-auth/private';
+import type { IoHelper } from '../io/private';
+import { IO } from '../io/private';
+
+/**
+ * Detect drift for a CloudFormation stack and wait for the detection to complete
+ *
+ * @param cfn - a CloudFormation client
+ * @param ioHelper - helper for IO operations
+ * @param stackName - the name of the stack to check for drift
+ * @returns the CloudFormation description of the drift detection results
+ */
+export async function detectStackDrift(
+  cfn: ICloudFormationClient,
+  ioHelper: IoHelper,
+  stackName: string,
+): Promise<DescribeStackResourceDriftsCommandOutput> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Starting drift detection for stack %s...', stackName)));
+
+  // Start drift detection
+  const driftDetection = await cfn.detectStackDrift({
+    StackName: stackName,
+  });
+
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Detecting drift for stack %s...', stackName)));
+
+  // Wait for drift detection to complete
+  const driftStatus = await waitForDriftDetection(cfn, ioHelper, driftDetection.StackDriftDetectionId!);
+
+  if (!driftStatus) {
+    throw new ToolkitError('Drift detection took too long to complete. Aborting');
+  }
+
+  if (driftStatus?.DetectionStatus === 'DETECTION_FAILED') {
+    throw new ToolkitError(
+      `Failed to detect drift: ${driftStatus.DetectionStatusReason || 'No reason provided'}`,
+    );
+  }
+
+  // Get the drift results
+  return cfn.describeStackResourceDrifts({
+    StackName: stackName,
+  });
+}
+
+/**
+ * Wait for a drift detection operation to complete
+ */
+async function waitForDriftDetection(
+  cfn: ICloudFormationClient,
+  ioHelper: IoHelper,
+  driftDetectionId: string,
+): Promise<DescribeStackDriftDetectionStatusCommandOutput | undefined> {
+  await ioHelper.notify(IO.CDK_TOOLKIT_I4506.msg(format('Waiting for drift detection %s to complete...', driftDetectionId)));
```

Comment:
Can we skip the initial waiting message? We just printed a message that we are starting the drift.

Reply:
Is the driftDetectionId useful here? I feel like this would be better as a single debug or trace message right after we get this.